### ISSUE TITLE
Hardcoded salt used for encryption

### ISSUE THREAD

**Issue by xtncl** [shlemiel/globaloe#2]
**Issue**
I noticed that the project uses a hardcoded salt for the `pbkdf2Async` function. While the implementation of PBKDF2 with a high number of iterations is commendable, the use of a static, hardcoded salt could potentially compromise the security, especially when two users have the same password. This setup results in identical encryption keys for identical passwords, which could make the system more vulnerable to certain types of cryptographic attacks.

**Generate unique salts**
To enhance the security posture of the project, I would like to suggest implementing unique salts for each encryption process. By generating a random salt for each encryption instance, we can ensure that even identical passwords will result in different encryption keys, significantly improving the system's resistance to attacks like rainbow table attacks.

**Storing the Salt**
Storing the Salt: The unique salt, along with the IV and ciphertext, can be safely stored together in the encrypted file. This practice is both secure and operationally convenient, as the salt does not need to be kept secret and is essential for the decryption process.

**Comment by shlemiel** [comment]
Thank you for the review.

I was aware of the issue, but honestly I doubted that there would be an adversary who will perform a rainbow table attack on our PBKDF2 salt value.

Besides, our adversary model assumes that the stored ciphertext only is accessible to attacker, so note that the overall security level depends on AES256-GCM, so sufficiently long and complex password will be a safeguard.

But you are right, I'll fix this. I am recently busy, so not so soon though... 😅

**Comment by xtncl** [comment]
You are right. This should be an easy fix and it would increase the security level by a lot, so i see no reason to not implement it :) Thanks!

**Comment by shlemiel** [comment]
Due to technical reason, this was very hard to implement.

In our design, PBKDF2 is implemented with `crypto.subtle` to cover up the performance issues. `crypto.subtle` provides only async cryptographic primitive functions, which greatly limited the possibility of per-markdown salt implementation. In detail, there were hooked callback functions to setup the newly open note data, which are `setViewData`, and `getViewData`, respectively. Unfortunately, they are related to internal  operations of the obsidian load/save mechanisms and cannot be changed to async.

Instead, added flexibility to input custom PBKDF2 salt in the setting tab.

**Comment by shlemiel** [comment]
For proof of concept, I tried to reimplement per-document based salt value, with asynchronous functions.

But, its user experience is not so smooth, since PBKDF2 creates a bit latency on every document opening. It is around a second on my personal computer, so...

1. If PBKDF2 iterations are reduced to reasonable level, one successful rainbow table attack on a document will lead to the password to be broken. Therefore, if compared to per-configuration custom salt value with large iterations, its security is weaker.
2. Finding reasonably fast and more secure new password derivation functions does not seem practically feasible, since current PBKDF2 implementation is mostly native to its execution environments (SubtleCrypto).
3. Even making selectable for salt value to be per-configuration based or per-document based is impossible, since they are not compatible to each other. It is more likely to be a migration.

Therefore, I postpone this suggestion.
